R/TealAppDriver.R

```diff
@@ -279,6 +280,22 @@ TealAppDriver <- R6::R6Class( # nolint: object_name.
           )
         )
       )
+
+      # Fallback if chrome/chromium doesn't support el.checkVisibility (< 105)
+      if (all(vapply(result, is.null, logical(1)))) {
+        fun_body <- "elem => !!( elem.offsetWidth || elem.offsetHeight || elem.getClientRects().length)"
+        unlist(
+          self$get_js(
+            sprintf(
+              "Array.from(document.querySelectorAll('%s')).map(%s)",
+              selector,
+              fun_body
+            )
+          )
+        )
+      } else {
+        result
+      }
```

Comment:
I really appreciate the finding here. Now I am wondering - is it better to provide the support for an old dependency or skip the test completely. The rationale here is that after upgrading (that everyone agrees it's needed) some part of code would leave not executed which I believe is a bad thing.
Taking this idea even further - we can introduce such requirements (I believe that this is not the only one) in the class constructor and stop very early before app init.
What do you think?

Reply:
I considered this and there's a solution to check for API support and skip: `typeof Element.prototype.checkVisibility === "function"` (js)

However, it didn't feel right to skip a test during integration tests, hence the fallback solution.

I'm fine with adding an automatic skip \*\*as long as\*\* the integration tests image is updated (see previous phrase :sweat_smile:)

>  Taking this idea even further - we can introduce such requirements (I believe that this is not the only one) in the class constructor and stop very early before app init.

This can be done with user agent (`navigator.userAgent` (js)). I didn't see a way of getting this info from `{chromote}` session, but maybe there is.

**By skipping I'm just worried about no actions being taken** if the image is reverted to a previous version or a new CI is introduced with an older version of chrome.

Failure here shows us that something is going on.

Reply:
My assumption here was that we will be looking into the test logs periodically and capture that some tests are unintentionally skipped. (That's actually the reason behind introducing non --as-cran pipeline but that's a separate story)
Thinking about this more - I think that skip would be more in line with testthat2 ~philosophy~ common practice. Please have a look at the purpose `skip_if_not_installed()` or any other `skip_if_xyz()` function - it is to do the skip (with a meaningful message) if some condition is met. That means that missing / insufficient version of test pre-requisites does not fail the tests. I consider testthat2, chromote and JS version as test pre-requisites. Note that we are not providing alternatives if they are missing.

Reply:
Ok, that sounds good as there is (will be) a plan in place to catch the skipped tests :+1: 

I'll try to find a way to get the versioning in the least `hacky` way possible. Hopefully, via `{chromote}` itself :sweat_smile:

Reply:
Added both a minimum version and a API-specific check.

I've added the general check to allow for future requirements that cannot be checked explicitly. But realistically, it could be removed for now.

### Minimum version

https://github.com/insightsengineering/teal/blob/09bfc1290dd6d089feabcb86e9ce264f8004704e/R/TealAppDriver.R#L70-L80

note that the version needs to be parsed

```r
# app <- TealAppDriver$new(...)
app$get_chromote_session()$Browser$getVersion()$product
#> [1] "HeadlessChrome/124.0.6367.60"
```

### API specific (less arbitrary and easier to maintain)

https://github.com/insightsengineering/teal/blob/09bfc1290dd6d089feabcb86e9ce264f8004704e/R/TealAppDriver.R#L287-L290

Reply:
awesome! thank you!